src/i2c_software.c

```diff
@@ -0,0 +1,164 @@
+// Software I2C emulation
+//
+// Copyright (C) 2023  Kevin O'Connor <kevin@koconnor.net>
+// Copyright (C) 2023  Alan.Ma <tech@biqu3d.com>
+//
+// This file may be distributed under the terms of the GNU GPLv3 license.
+
+#include <string.h> // memcpy
+#include "board/gpio.h" // gpio_out_setup
+#include "board/internal.h" // gpio_peripheral
+#include "board/misc.h" // timer_read_time
+#include "basecmd.h" // oid_alloc
+#include "command.h" // DECL_COMMAND
+#include "sched.h" // sched_shutdown
+#include "i2ccmds.h" // i2cdev_set_software_bus
+
+struct i2c_software {
+    struct gpio_out scl, sda;
+    uint32_t sda_pin;
+    uint8_t addr;
+    unsigned int ticks;
+};
+
+void
+command_i2c_set_software_bus(uint32_t *args)
+{
+    struct i2cdev_s *i2c = i2cdev_oid_lookup(args[0]);
+    struct i2c_software *is = alloc_chunk(sizeof(*is));
+    // Max speed = 1MHz
+    unsigned int speed = args[3] > 1000000 ? 1000000 : args[3];
+    is->ticks = 1000000000 / speed / 2;
+    is->addr = (args[4] & 0x7f) << 1; // address format shifted
+    is->sda_pin = args[2];
+    is->scl = gpio_out_setup(args[1], 1);
+    is->sda = gpio_out_setup(is->sda_pin, 1);
+    i2cdev_set_software_bus(i2c, is);
+}
+DECL_COMMAND(command_i2c_set_software_bus,
+             "i2c_set_software_bus oid=%c scl_pin=%u sda_pin=%u"
+             " rate=%u address=%u");
+
+static unsigned int
+nsecs_to_ticks(uint32_t ns)
+{
+    return timer_from_us(ns * 1000) / 1000000;
+}
+
+void i2c_delay(unsigned int n) {
+    unsigned int t = timer_read_time() + nsecs_to_ticks(n);
+    while (t > timer_read_time());
+}
+
+static void
+i2c_software_send_ack(struct i2c_software *is, const uint8_t ack)
+{
+    gpio_out_write(is->sda, ack);
+    i2c_delay(is->ticks);
+    gpio_out_write(is->scl, 1);
+    i2c_delay(is->ticks);
+    gpio_out_write(is->scl, 0);
+}
+
+static uint8_t
+i2c_software_read_ack(struct i2c_software *is)
+{
+    uint8_t nack = 0;
+    struct gpio_in sda_in = gpio_in_setup(is->sda_pin, 1);
+    i2c_delay(is->ticks);
+    gpio_out_write(is->scl, 1);
+    nack = gpio_in_read(sda_in);
+    i2c_delay(is->ticks);
+    gpio_out_write(is->scl, 0);
+    gpio_out_setup(is->sda_pin, 1);
+    return nack;
+}
+
+static void
+i2c_software_send_byte(struct i2c_software *is, uint8_t b)
+{
+    for (uint_fast8_t i = 0; i < 8; i++) {
+        gpio_out_write(is->sda, b & 0x80);
+        b <<= 1;
+        i2c_delay(is->ticks);
+        gpio_out_write(is->scl, 1);
+        i2c_delay(is->ticks);
+        gpio_out_write(is->scl, 0);
```

Comment:
To communicate over an I2C bus, the transmitter should transition between GND and high-Z states.  This code is transitioning between GND and HIGH states.  If the transmitter transitions to a HIGH state while the receiver is in a GND state it would result in a "short" on the wires that would consume power, generate heat, and possibly destabilize the board voltage.

I have written code to communicate with I2C devices using just GND/HIGH transitions (eg, `klippy/extras/mcp4018.py`).  This change, though, may confuse users into thinking Klipper supports I2C on any device.  I'm not sure that is a good idea.

Reply:
Yeah, Theoretically this is a problem.
It would be perfect if there were a universal API to set GPIO as Open-Drain output, but some MCU (such as RP2040) do not seem to have the Open-Drain feature

Reply:
Could the code transmit bits by switching between `gpio_in_reset(is->sda_in, 1)` and `gpio_out_reset(is->sda_out, 0)`?

I suspect the transfer rate would be slower, but it would hopefully not cause electrical problems.

Reply:
Yeah, the transfer rate would be slower. Mainly affecting MCU with lower main frequencies such as STM32G0B1 64MHz. 
The measured results are as follows

1 | `gpio_out_write` | `gpio_in/out_reset`
-- | -- | --
STM32G0B1 | 70KHz  | 30-40KHz
RP2040 | 84KHz  | 70KHz

For MCUs with high main frequencies like `STM32H7`,  The impact is minimal.